pep-0624.rst

```diff
@@ -0,0 +1,233 @@
+improves the maintainability of CPython, but it also helps API users
+and other Python implementations.
+
+
+Rationale
+=========
+
+Deprecated since Python 3.3
+---------------------------
+
+``Py_UNICODE`` and APIs using it are deprecated since Python 3.3.
+
+
+Inefficient
+-----------
+
+All of these APIs are implemented using ``PyUnicode_FromWideChar``.
+So these APIs are inefficient when user want to encode Unicode
+object.
+
+
+Not used widely
+---------------
+
+When searching from top 4000 PyPI packages [1]_, only pyodbc use
+these APIs.
+
+* ``PyUnicode_EncodeUTF8()``
+* ``PyUnicode_EncodeUTF16()``
+
+pyodbc uses these APIs to encode Unicode object into bytes object.
+So it is easy to fix it. [2]_
+
+
+Alternative APIs
+================
+
+There are alternative APIs to accept ``PyObject *unicode`` instead of
+``Py_UNICODE *``.
+
+
+========================================= ====================================================================
+Deprecated API                            Alternative APIs
+========================================= ====================================================================
+``PyUnicode_Encode()``                    ``PyUnicode_AsEncodedString()``
+``PyUnicode_EncodeASCII()``               ``PyUnicode_AsASCIIString()`` \(1), ``_PyUnicode_AsASCIIString()``
+``PyUnicode_EncodeLatin1()``              ``PyUnicode_AsLatin1String()`` \(1), ``_PyUnicode_AsLatin1String()``
+``PyUnicode_EncodeUTF7()``                ``_PyUnicode_EncodeUTF7()``
+``PyUnicode_EncodeUTF8()``                ``PyUnicode_AsUTF8String()`` \(1), ``_PyUnicode_AsUTF8String()``
+``PyUnicode_EncodeUTF16()``               ``PyUnicode_AsUTF16String()`` \(2), ``_PyUnicode_EncodeUTF16()``
+``PyUnicode_EncodeUTF32()``               ``PyUnicode_AsUTF32String()`` \(2), ``_PyUnicode_EncodeUTF32()``
+``PyUnicode_EncodeUnicodeEscape()``       ``PyUnicode_AsUnicodeEscapeString()``
+``PyUnicode_EncodeRawUnicodeEscape()``    ``PyUnicode_AsRawUnicodeEscapeString()``
+``PyUnicode_EncodeCharmap()``             ``PyUnicode_AsCharmapString()`` \(1)
+``PyUnicode_TranslateCharmap()``          ``PyUnicode_Translate()``
+``PyUnicode_EncodeDecimal()``              \(3)
+``PyUnicode_TransformDecimalToASCII()``    \(3)
+========================================= ====================================================================
+
+Notes:
+
+(1)
+   ``const char *errors`` is missing.
+
+(2)
+   ``const char *errors, int byteorder`` are missing.
+
+(3)
+   There is no direct replacement. But CPython uses
+   ``_PyUnicode_TransformDecimalAndSpaceToASCII``
+   for converting from Unicode to numbers instead.
+
+
+Plan
+====
+
+Python 3.9
+----------
+
+Add ``Py_DEPRECATED(3.3)`` to these APIs.
+
+* ``PyUnicode_EncodeDecimal()``
+* ``PyUnicode_TransformDecimalToASCII()``.
+
+This change is committed already.
+https://github.com/python/cpython/commit/9c3840870814493fed62e140cfa43c2883e12181
+
+All other APIs have been marked ``Py_DEPRECATED(3.3)`` already.
```

Comment:
Currently, the documentation says ".. deprecated-removed:: 3.3 4.0". Can you please also update this documentation, and announce the removal in What's New in Python 3.9?

Reply:
I will update them after this PEP is accepted, because this PEP is written there is no consensus when it is removed.

> I'm not comfortable with removing functions without emitting warning **at runtime** during at least one release. Maybe:

I don't like runtime warning.  For example, `PyUnicode_EncodeUTF8(PyUnicode_AS_UNICODE(obj), ...)` may cause SEGV if obj is borrowed reference.